codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1047 @@
+                        }
+                    }
+
+                    pipeline().fireChannelInactive();
+                    if (isRegistered()) {
+                        deregister(unsafe().voidPromise());
+                    }
+                    promise.setSuccess();
+                    closePromise.setSuccess();
+                } finally {
+                    inClose = false;
+                }
+            }
+
+            @Override
+            public void closeForcibly() {
+                close(unsafe().voidPromise());
+            }
+
+            @Override
+            public void deregister(ChannelPromise promise) {
+                if (registered) {
+                    registered = true;
+                    promise.setSuccess();
+                    pipeline().fireChannelUnregistered();
+                } else {
+                    promise.setFailure(new IllegalStateException("Not registered"));
+                }
+            }
+
+            @Override
+            public void beginRead() {
+                if (readInProgress || !isActive()) {
+                    return;
+                }
+
+                final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+                allocHandle.reset(config());
+                if (inboundBuffer == null) {
+                    if (closePending) {
+                        unsafe.closeForcibly();
+                    } else {
+                        readInProgress = true;
+                    }
+                    return;
+                }
+
+                do {
+                    Object m = inboundBuffer.poll();
+                    if (m == null && closePending) {
+                        pipeline().fireChannelReadComplete();
+                        unsafe.closeForcibly();
+                        return;
+                    }
+                    doRead0((Http2Frame) m, allocHandle);
+                } while (allocHandle.continueReading());
+
+                allocHandle.readComplete();
+                pipeline().fireChannelReadComplete();
+            }
+
+            /**
+             * Returns whether reads should continue. The only reason reads shouldn't continue is that the
+             * channel was just closed.
+             */
+            @SuppressWarnings("deprecation")
+            void doRead0(Http2Frame frame, RecvByteBufAllocator.Handle allocHandle) {
+                int numBytesToBeConsumed = 0;
+                if (frame instanceof Http2DataFrame) {
+                    numBytesToBeConsumed = ((Http2DataFrame) frame).flowControlledBytes();
+                    allocHandle.lastBytesRead(numBytesToBeConsumed);
+                } else {
+                    allocHandle.lastBytesRead(ARBITRARY_MESSAGE_SIZE);
+                }
+                allocHandle.incMessagesRead(1);
+                pipeline().fireChannelRead(frame);
+
+                if (numBytesToBeConsumed != 0) {
```

Comment:
we should investigate a means to provide bytes to flow control asynchronously in the future https://github.com/netty/netty/issues/7078.